### ISSUE TITLE
resource add: "unknown_error: Something went wrong"

### ISSUE THREAD

**Issue by th3m4ri0** [flynn/flynn#2257]
Can't create a postgres database :anguished: 
```
$ flynn resource add postgres
unknown_error: Something went wrong
```

**Comment by lmars** [comment]
@th3m4ri0 please grab the controller web log with `flynn -a controller log --process-type web --raw-output` and post / gist the output

**Comment by th3m4ri0** [comment]
```
t=2015-12-08T02:52:13+0000 lvl=eror msg="validation_error: must either request a stream or only active formations" component=controller req_id=4e3d9aa5-535c-4a50-9388-d177e276aa99
t=2015-12-08T02:52:13+0000 lvl=info msg="request completed" component=controller req_id=4e3d9aa5-535c-4a50-9388-d177e276aa99 status=200 duration=13.203665ms
t=2015-12-08T02:53:13+0000 lvl=info msg="request started" component=controller req_id=e5001255-b667-44e0-b692-021bd2b4dcae method=GET path=/formations client_ip=REDACTED
t=2015-12-08T02:53:13+0000 lvl=eror msg="validation_error: must either request a stream or only active formations" component=controller req_id=e5001255-b667-44e0-b692-021bd2b4dcae
t=2015-12-08T02:53:13+0000 lvl=info msg="request completed" component=controller req_id=e5001255-b667-44e0-b692-021bd2b4dcae status=200 duration=11.815462ms
t=2015-12-08T02:53:37+0000 lvl=info msg="request started" component=controller req_id=8c8d3432-ac41-440d-8866-ad2c3446c084 method=POST path=/providers/postgres/resources client_ip=REDACTED
t=2015-12-08T02:53:37+0000 lvl=eror msg="ERROR: more than one row returned by a subquery used as an expression (SQLSTATE 21000)" component=controller req_id=8c8d3432-ac41-440d-8866-ad2c3446c084
t=2015-12-08T02:53:37+0000 lvl=info msg="request completed" component=controller req_id=8c8d3432-ac41-440d-8866-ad2c3446c084 status=500 duration=811.956384ms
t=2015-12-08T02:53:38+0000 lvl=info msg="request started" component=controller req_id=5364da15-e953-4188-98d2-0553b65739af method=POST path=/providers/postgres/resources client_ip=REDACTED
t=2015-12-08T02:53:39+0000 lvl=eror msg="ERROR: more than one row returned by a subquery used as an expression (SQLSTATE 21000)" component=controller req_id=5364da15-e953-4188-98d2-0553b65739af
t=2015-12-08T02:53:39+0000 lvl=info msg="request completed" component=controller req_id=5364da15-e953-4188-98d2-0553b65739af status=500 duration=734.227363ms
t=2015-12-08T02:53:39+0000 lvl=info msg="request started" req_id=fd3f7285-32a6-4938-9bab-f0d2f3656e40 component=controller method=POST path=/providers/postgres/resources client_ip=REDACTED
t=2015-12-08T02:53:40+0000 lvl=eror msg="ERROR: more than one row returned by a subquery used as an expression (SQLSTATE 21000)" req_id=fd3f7285-32a6-4938-9bab-f0d2f3656e40 component=controller
t=2015-12-08T02:53:40+0000 lvl=info msg="request completed" req_id=fd3f7285-32a6-4938-9bab-f0d2f3656e40 component=controller status=500 duration=784.096543ms
t=2015-12-08T02:53:40+0000 lvl=info msg="request started" component=controller req_id=067a96eb-b497-4df6-92c7-36ae2fbe09bc method=POST path=/providers/postgres/resources client_ip=REDACTED
t=2015-12-08T02:53:40+0000 lvl=eror msg="ERROR: more than one row returned by a subquery used as an expression (SQLSTATE 21000)" component=controller req_id=067a96eb-b497-4df6-92c7-36ae2fbe09bc
t=2015-12-08T02:53:40+0000 lvl=info msg="request completed" component=controller req_id=067a96eb-b497-4df6-92c7-36ae2fbe09bc status=500 duration=421.853979ms
t=2015-12-08T02:53:41+0000 lvl=info msg="request started" component=controller req_id=3848d60a-52a6-4c47-a6bd-5d5c7dd117ed method=POST path=/providers/postgres/resources client_ip=REDACTED
t=2015-12-08T02:53:41+0000 lvl=eror msg="ERROR: more than one row returned by a subquery used as an expression (SQLSTATE 21000)" component=controller req_id=3848d60a-52a6-4c47-a6bd-5d5c7dd117ed
t=2015-12-08T02:53:41+0000 lvl=info msg="request completed" component=controller req_id=3848d60a-52a6-4c47-a6bd-5d5c7dd117ed status=500 duration=746.855056ms
t=2015-12-08T02:53:42+0000 lvl=info msg="request started" component=controller req_id=fc340941-3c7c-4429-be77-4909de625c50 method=POST path=/providers/postgres/resources client_ip=REDACTED
t=2015-12-08T02:53:43+0000 lvl=eror msg="ERROR: more than one row returned by a subquery used as an expression (SQLSTATE 21000)" component=controller req_id=fc340941-3c7c-4429-be77-4909de625c50
t=2015-12-08T02:53:43+0000 lvl=info msg="request completed" component=controller req_id=fc340941-3c7c-4429-be77-4909de625c50 status=500 duration=1.050378924s
t=2015-12-08T02:53:43+0000 lvl=info msg="request started" component=controller req_id=5ab68bdf-fed8-4189-961e-a39bad3b11bf method=POST path=/providers/postgres/resources client_ip=REDACTED
t=2015-12-08T02:53:44+0000 lvl=eror msg="ERROR: more than one row returned by a subquery used as an expression (SQLSTATE 21000)" component=controller req_id=5ab68bdf-fed8-4189-961e-a39bad3b11bf
t=2015-12-08T02:53:44+0000 lvl=info msg="request completed" component=controller req_id=5ab68bdf-fed8-4189-961e-a39bad3b11bf status=500 duration=788.303769ms
t=2015-12-08T02:53:44+0000 lvl=info msg="request started" component=controller req_id=4d2bf712-99e9-41bd-a8ae-ca3a44c532b5 method=POST path=/providers/postgres/resources client_ip=REDACTED
t=2015-12-08T02:53:45+0000 lvl=eror msg="ERROR: more than one row returned by a subquery used as an expression (SQLSTATE 21000)" component=controller req_id=4d2bf712-99e9-41bd-a8ae-ca3a44c532b5
t=2015-12-08T02:53:45+0000 lvl=info msg="request completed" component=controller req_id=4d2bf712-99e9-41bd-a8ae-ca3a44c532b5 status=500 duration=872.313997ms
t=2015-12-08T02:53:45+0000 lvl=info msg="request started" component=controller req_id=bb2706f6-2e36-41a2-be55-41df3df522e0 method=POST path=/providers/postgres/resources client_ip=REDACTED
t=2015-12-08T02:53:46+0000 lvl=eror msg="ERROR: more than one row returned by a subquery used as an expression (SQLSTATE 21000)" component=controller req_id=bb2706f6-2e36-41a2-be55-41df3df522e0
t=2015-12-08T02:53:46+0000 lvl=info msg="request completed" component=controller req_id=bb2706f6-2e36-41a2-be55-41df3df522e0 status=500 duration=657.813641ms
t=2015-12-08T02:53:46+0000 lvl=info msg="request started" component=controller req_id=c001eba4-8c41-4d43-abf2-76b8b0f02a99 method=POST path=/providers/postgres/resources client_ip=REDACTED
t=2015-12-08T02:53:47+0000 lvl=eror msg="ERROR: more than one row returned by a subquery used as an expression (SQLSTATE 21000)" component=controller req_id=c001eba4-8c41-4d43-abf2-76b8b0f02a99
t=2015-12-08T02:53:47+0000 lvl=info msg="request completed" component=controller req_id=c001eba4-8c41-4d43-abf2-76b8b0f02a99 status=500 duration=1.196991849s
t=2015-12-08T02:54:13+0000 lvl=info msg="request started" req_id=d0376eca-61e3-4594-ad9e-9944cbba683a component=controller method=GET path=/formations client_ip=REDACTED
t=2015-12-08T02:54:13+0000 lvl=eror msg="validation_error: must either request a stream or only active formations" req_id=d0376eca-61e3-4594-ad9e-9944cbba683a component=controller
t=2015-12-08T02:54:13+0000 lvl=info msg="request completed" req_id=d0376eca-61e3-4594-ad9e-9944cbba683a component=controller status=200 duration=7.894873ms
t=2015-12-08T02:54:21+0000 lvl=info msg="request started" component=controller req_id=e8564bf9-fc9e-441f-a67f-f8fecd6e97ef method=GET path=/apps/controller/log client_ip=REDACTED
```

**Comment by lmars** [comment]
@josephglanville any thoughts on that postgres error?

**Comment by josephglanville** [comment]
@lmars th

**Comment by josephglanville** [comment]
@lmars the issue is definitely one of these 2 queries `app_resource_insert_app_by_name`, `app_resource_insert_app_by_name_or_id`. I will take a closer look at the constraints etc and see if I can work out why it seems like there could be more than 1 record returned by either of those sub queries that select from the apps table.

@th3m4ri0 is it possible you could dump the apps table and gist it? Or even just the `app_id` and `name` columns which could be done like this:
```
flynn -a controller pg psql -- -c 'SELECT name, app_id from apps'
```

**Comment by josephglanville** [comment]
So this is indeed puzzling. The `app_id` column is `PRIMARY KEY` and the `name` column has a `UNIQUE` index on it... so it shouldn't be possible for either of them to collide. However, there is a chance that if you created an app with a UUID value for `name` that already exists in the `app_id` column you could cause this error.

**Comment by th3m4ri0** [comment]
```
flynn -a controller pg psql -- -c 'SELECT name, app_id from apps'
          name           |                app_id
-------------------------+--------------------------------------
 controller              | 51fb1bb6-e9f6-4b46-a723-1fed5cdd2341
 postgres                | 66f103fc-97bc-46c4-a6d8-ae86ab9ec39e
 flannel                 | 96df53a3-acbf-4609-b9fa-e7e40ccf3c65
 discoverd               | ad74073d-bbd0-4d5f-a76e-eb1465ec2106
 blobstore               | 72a3774d-caf2-4762-a58a-3c44f541e68e
 router                  | 4cb08094-7717-402b-ba89-d537fa18e419
 gitreceive              | 1db6e4d1-99c8-418f-b0bd-162ce0097e1c
 logaggregator           | 069952ed-b348-4489-a5f7-16aa93ce1cd1
 taffy                   | cc3a5019-7224-4f23-96ee-04a29b7e8fb9
 dashboard               | 877ed098-681d-4489-b465-ecf108f89b88
 status                  | 8aae4d0a-e2b4-4512-b77c-b68240528bc6
 example                 | 62a6b5e4-d15f-42ce-a713-b2a71782e605
 swallows-disarm-kearney | 8793a0a0-f43a-4d82-b411-d50d127acf3b
 test-db                    | 9f0ec23e-9660-497a-8c23-31e40c5b8ddd
 test-db                    | 103fb71c-c565-475d-a67d-cf7558f9f263
 example-staging         | ffd3af98-e4de-4c00-8bb9-ae496e3ab361
 example                 | 84e6c229-eb98-4764-869d-37d3b155125e
```

I guess test-db is here twice because I deleted then recreated the app ?

**Comment by titanous** [comment]
Yeah, that's the problem, patched in #2279.

**Comment by th3m4ri0** [comment]
Can't wait ! :raised_hands:
